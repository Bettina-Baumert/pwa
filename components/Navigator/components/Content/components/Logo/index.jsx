/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { logo } from 'Config/app.json';
import styles from './style';

/**
 * The navigator logo component.
 * @return {JSX}
 */
const Logo = () => <img className={styles} src={logo} alt="Logo" />;

export default Logo;
