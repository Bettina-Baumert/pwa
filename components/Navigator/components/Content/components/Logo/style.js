import { css } from 'glamor';

export default css({
  margin: '0 auto',
  maxHeight: '100%',
});
