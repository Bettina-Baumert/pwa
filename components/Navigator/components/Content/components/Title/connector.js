import { connect } from 'react-redux';
import { getTitle } from 'Components/View/selectors';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  title: getTitle(state),
});

/**
 * @param {Object} next The next state.
 * @param {Object} prev The previous state.
 * @returns {boolean}
 */
const areStatePropsEqual = (next, prev) => (
  prev.title === next.title
);

export default connect(mapStateToProps, null, null, { areStatePropsEqual });
