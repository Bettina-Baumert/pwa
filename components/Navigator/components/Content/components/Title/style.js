/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import cxs from 'cxs';
import variables from 'Styles/variables';

/**
 * Styling for the title.
 */
const title = cxs({
  position: 'absolute',
  top: '.65em',
  left: 0,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  width: '100%',
  padding: `0 ${variables.gap.big}px`,
});

/**
 * Animation configuration
 */
const distance = '20px';
const duration = 300; // The duration of the animation in ms.
const center = {
  opacity: 1,
  transform: 'translate3d(0, 0, 0)',
};
const right = {
  opacity: 0,
  transform: `translate3d(${distance}, 0, 0)`,
};
const left = {
  opacity: 0,
  transform: `translate3d(-${distance}, 0, 0)`,
};

/**
 * Keyframe animation for all title positions.
 */
cxs({
  '@keyframes navTitleCenterToRight': {
    from: center,
    to: right,
  },
  '@keyframes navTitleCenterToLeft': {
    from: center,
    to: left,
  },
  '@keyframes navTitleRightToCenter': {
    from: right,
    to: center,
  },
  '@keyframes navTitleLeftToCenter': {
    from: left,
    to: center,
  },
});

/**
 * Creates a simple class that is using a keyframe
 * @param {string} name The name of the keyframe
 * @returns {string} Class name
 */
const createKeyframeAnimationClass = name => cxs({
  animation: `navTitle${name} ${duration}ms cubic-bezier(0.25, 1, 0.25, 1)`,
});

const hidden = cxs({
  opacity: 0,
});

export default {
  title,
  hidden,
  centerToRight: createKeyframeAnimationClass('CenterToRight'),
  centerToLeft: createKeyframeAnimationClass('CenterToLeft'),
  rightToCenter: createKeyframeAnimationClass('RightToCenter'),
  leftToCenter: createKeyframeAnimationClass('LeftToCenter'),
};
