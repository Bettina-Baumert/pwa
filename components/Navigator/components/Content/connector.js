import { connect } from 'react-redux';
import { getCurrentPathname, getCurrentSearchQuery } from '@shopgate/pwa-common/selectors/router';
import submitSearch from '../../actions/submitSearch';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  path: getCurrentPathname(state),
  searchQuery: getCurrentSearchQuery(state),
  searchActive: state.navigator.searchActive,
});

/**
 * Connects the dispatch function to a callable function in the props.
 * @param {Function} dispatch The redux dispatch function.
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  submitSearch: () => dispatch(submitSearch()),
});

/**
 * @param {Object} next The next component props.
 * @param {Object} prev The previous component props.
 * @return {boolean}
 */
const areStatePropsEqual = (next, prev) => {
  if ((!prev.path && next.path) || (prev.path !== next.path)) {
    return false;
  }

  if ((!prev.searchQuery && next.searchQuery) || (prev.searchQuery !== next.searchQuery)) {
    return false;
  }

  if (prev.searchActive !== next.searchActive) {
    return false;
  }

  return true;
};

export default connect(mapStateToProps, mapDispatchToProps, null, { areStatePropsEqual });
