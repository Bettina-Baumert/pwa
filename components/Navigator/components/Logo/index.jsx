import React, { Component, PropTypes } from 'react';
import connect from '../../connector';
import styles from './style';

/**
 * The navigator logo component.
 */
class Logo extends Component {
  static propTypes = {
    logoUrl: PropTypes.string,
  };

  static defaultProps = {
    logoUrl: null,
  };

  /**
   * Constructor
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.isLoaded = false;

    this.state = {
      url: this.props.logoUrl,
    };
  }

  /**
   * The componentDidMount lifecycle hook.
   * Render the image if we have the url in the props.
   */
  componentDidMount() {
    if (this.props.logoUrl) {
      this.loadImage(this.props.logoUrl);
    }
  }

  /**
   * The componentWillReceiveProps lifecycle hook.
   * @param {Object} nextProps The next set of props.
   */
  componentWillReceiveProps(nextProps) {
    if (!this.isLoaded) {
      this.loadImage(nextProps.logoUrl);
    }
  }

  /**
   * The render function.
   * @param {Object} nextProps The next set of props.
   * @returns {boolean}
   */
  shouldComponentUpdate() {
    return !this.isLoaded;
  }

  /**
   * Updates the state with the given url.
   * @param {string} url The image url.
   */
  setUrl = (url) => {
    this.loaded = true;
    this.setState({ url });
  }

  /**
   * Loads an image with the given url.
   * @param {string} url The image url.
   * @returns {Function}
   */
  loadImage = (url) => {
    const image = new Image();
    image.src = url;
    image.onload = () => this.setUrl(url);

    return image.complete;
  }

  /**
   * The render function.
   * @returns {JSX}
   */
  render() {
    if (!this.state.url) {
      return null;
    }

    return <img className={styles.logo} src={this.props.logoUrl} alt="Logo" />;
  }
}

export default connect(Logo);
