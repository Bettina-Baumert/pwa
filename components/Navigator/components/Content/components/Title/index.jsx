/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { HISTORY_POP_ACTION } from '@shopgate/pwa-common/constants/History';
import { KeyframeAnimation } from '@shopgate/pwa-common/components/KeyframeAnimation';
import connect from '../../../../connector';
import styles from './style';

/**
 * The Navigator Title component.
 */
class Title extends Component {
  static propTypes = {
    action: PropTypes.string.isRequired,
    historyLength: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    onClick: PropTypes.func,
  };

  static defaultProps = {
    onClick: () => {},
  };

  /**
   * The constructor.
   * Sets the initial state from connected props.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.previousTitle = null;
    this.title = props.title;
  }

  /**
   * Checks if the title was reset and if the component should update.
   * @param {Object} nextProps The next incoming props.
   * @returns {boolean} Whether the component should update.
   */
  shouldComponentUpdate(nextProps) {
    return (this.props.title !== nextProps.title) && nextProps.title;
  }

  /**
   * When the component updates, preserve the current title.
   * @param {Object} nextProps The components next props.
   */
  componentWillUpdate(nextProps) {
    this.previousTitle = this.title;
    this.title = nextProps.title;
  }

  /**
   * Determine how the views should be displayed.
   * @returns {JSX}
   */
  get transitionClass() {
    // For the first page don't do any animation.
    if (this.props.historyLength === 1) {
      return {
        inactive: '',
        active: '',
      };
    }

    const pop = this.props.action === HISTORY_POP_ACTION;

    return {
      inactive: pop ? styles.centerToRight : styles.centerToLeft,
      active: pop ? styles.leftToCenter : styles.rightToCenter,
    };
  }

  /**
   * Renders the component.
   * Displays the Title.
   * @returns {JSX}
   */
  render() {
    return (
      <div aria-hidden onClick={this.props.onClick}>
        {/* Renders the inactive / previous title */}
        <KeyframeAnimation
          className={styles.title}
          activeClass={this.transitionClass.inactive}
          inactiveClass={styles.hidden}
        >
          {this.previousTitle}
        </KeyframeAnimation>

        {/* Renders the active / current title */}
        <KeyframeAnimation
          className={styles.title}
          activeClass={this.transitionClass.active}
        >
          {this.props.title}
        </KeyframeAnimation>
      </div>
    );
  }
}

export default connect(Title);
