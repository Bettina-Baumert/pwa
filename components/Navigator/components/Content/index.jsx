/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { INDEX_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import connect from './connector';
import Logo from './components/Logo';
import Title from './components/Title';

/**
 * The navigator content component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Content = (props) => {
  if (props.path === INDEX_PATH) {
    return <Logo />;
  }

  return <Title />;
};

Content.propTypes = {
  path: PropTypes.string.isRequired,
};

export default connect(Content);
