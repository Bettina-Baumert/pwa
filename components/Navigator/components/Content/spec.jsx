import React from 'react';
import { shallow } from 'enzyme';
import Content from './index';
import Logo from '../Logo';
import Title from '../Title';

// Mock the redux connect() method instead of providing a fake store.
jest.mock('Library/connectors/navigator', () => (obj) => {
  const newObj = obj;

  newObj.defaultProps = {
    ...newObj.defaultProps,
    historyLength: 1,
    action: 'POP',
    title: 'My Title',
    submitSearch: () => {},
  };

  return newObj;
});

// Mock the redux connect() method instead of providing a fake store.
jest.mock('Library/connectors/history', () => (obj) => {
  const newObj = obj;
  newObj.defaultProps = {
    ...newObj.defaultProps,
    getQueryParam: () => null,
  };
  return newObj;
});

describe('<Content />', () => {
  it('should render the logo', () => {
    const wrapper = shallow(
      <Content path="/" />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Logo).exists()).toBe(true);
  });

  it('should render the title', () => {
    const wrapper = shallow(
      <Content path="some/other/path" />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Title).exists()).toBe(true);
  });
});
