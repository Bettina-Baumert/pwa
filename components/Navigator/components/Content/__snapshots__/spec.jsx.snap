// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigatorContent /> should render the logo 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NavigatorContent
    routePattern="/"
  >
    <Logo>
      <img
        alt="Shopgate Connect"
        className={
          Object {
            "data-css-bqj31d": "",
          }
        }
        src="https://example.com/logo"
      />
    </Logo>
    <_default>
      <Consumer>
        <NavigatorSearch
          active={false}
          query=""
          setSearchQuery={[MockFunction]}
          toggleSearchField={[MockFunction]}
        >
          <Transition
            appear={false}
            enter={true}
            exit={true}
            in={false}
            mountOnEnter={false}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            timeout={150}
            unmountOnExit={false}
          >
            <form
              className="css-15xpwm2"
              data-test-id="Search"
              onSubmit={[Function]}
              style={
                Object {
                  "transform": "translate3d(100vw, 0, 0)",
                }
              }
            >
              <input
                className="css-1obolt5"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="search.placeholder"
                type="search"
                value=""
              />
              <div
                aria-hidden={true}
                className="css-lvd6ix"
                onClick={[Function]}
                role="button"
              />
            </form>
          </Transition>
        </NavigatorSearch>
      </Consumer>
    </_default>
  </NavigatorContent>
</Provider>
`;

exports[`<NavigatorContent /> should render the title 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NavigatorContent
    routePattern="some/other/path"
  >
    <Connect(Title)>
      <Title
        dispatch={[Function]}
        onClick={[Function]}
        title="test"
      >
        <div
          aria-hidden={true}
          onClick={[Function]}
        >
          <div
            className="css-iykibm "
          >
            <Translate
              className=""
              params={Object {}}
              string=""
            />
          </div>
          <div
            className="css-iykibm "
            data-test-id="title: test"
          >
            <Translate
              className=""
              params={Object {}}
              string="test"
            >
              <span
                className=""
              >
                test
              </span>
            </Translate>
          </div>
        </div>
      </Title>
    </Connect(Title)>
    <_default>
      <Consumer>
        <NavigatorSearch
          active={false}
          query=""
          setSearchQuery={[MockFunction]}
          toggleSearchField={[MockFunction]}
        >
          <Transition
            appear={false}
            enter={true}
            exit={true}
            in={false}
            mountOnEnter={false}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            timeout={150}
            unmountOnExit={false}
          >
            <form
              className="css-15xpwm2"
              data-test-id="Search"
              onSubmit={[Function]}
              style={
                Object {
                  "transform": "translate3d(100vw, 0, 0)",
                }
              }
            >
              <input
                className="css-1obolt5"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="search.placeholder"
                type="search"
                value=""
              />
              <div
                aria-hidden={true}
                className="css-lvd6ix"
                onClick={[Function]}
                role="button"
              />
            </form>
          </Transition>
        </NavigatorSearch>
      </Consumer>
    </_default>
  </NavigatorContent>
</Provider>
`;
